presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveDistinctFromSemiJoin.java

```diff
@@ -0,0 +1,98 @@
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+/**
+ * Remove distinct node from following path:
+ * <pre>
+ *     SemiJoinNode
+ *        - Source
+ *        - FilteringSource
+ *          - ProjectNode
+ *              - AggregationNode
+ * </pre>
+ */
+public class RemoveDistinctFromSemiJoin
+    implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (node instanceof SemiJoinNode && session.getSystemProperty("remove_distinct_from_semijoin", Boolean.class)) {
+            return rewrite(node, lookup);
+        }
+        return Optional.empty();
+    }
+
+   private Optional<PlanNode> rewrite(PlanNode node, Lookup lookup)
+   {
+        if (node instanceof SemiJoinNode) {
+            PlanNode filteringSource = lookup.resolve(((SemiJoinNode) node).getFilteringSource());
+            if (filteringSource instanceof AggregationNode) {
+                AggregationNode aggregationNode = (AggregationNode) filteringSource;
+                if (isDistinctNode(aggregationNode)) {
+                    return Optional.of(
+                            node.replaceChildren(ImmutableList.of(
+                                    ((SemiJoinNode) node).getSource(), aggregationNode.getSource())
+                            )
+                    );
+                }
+            }
+            else if (filteringSource instanceof  ProjectNode) {
+                Optional<PlanNode> result = rewrite(filteringSource, lookup);
+                if (result.isPresent()) {
+                    return Optional.of(node.replaceChildren(ImmutableList.of(((SemiJoinNode) node).getSource(), result.get())));
+                }
+            }
+        }
+
+        if (node instanceof ProjectNode) {
+            PlanNode child = lookup.resolve(((ProjectNode) node).getSource());
+            if (child instanceof AggregationNode) {
+                AggregationNode aggregationNode = (AggregationNode) child;
+                if (isDistinctNode(aggregationNode)) {
+                    return Optional.of(node.replaceChildren(ImmutableList.of(aggregationNode.getSource())));
+                }
+            }
+            else {
+                Optional<PlanNode> result = rewrite(((ProjectNode) node).getSource(), lookup);
+                if (result.isPresent()) {
+                    return Optional.of(node.replaceChildren(ImmutableList.of(result.get())));
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private boolean isDistinctNode(AggregationNode aggregationNode)
```

Comment:
You should probably remove only `aggregationNode.getStep() == SINGLE`. I.e. this rule should fire when there are no PARTIAL aggregations in the plan yet, because it doesn't handle them.